shared/pubsub.js

```diff
@@ -0,0 +1,397 @@
+  timeout: 5000
+}
+
+const eventNames = ['close', 'error', 'message', 'open']
+const globalEventNames = ['offline', 'online']
+
+// Parses and validates a received message.
+export const messageParser = (data: string): Message => {
+  const msg = JSON.parse(data)
+
+  if (typeof msg !== 'object' || msg === null) {
+    throw new TypeError('Message is null or not an object')
+  }
+  const { type } = msg
+
+  if (typeof type !== 'string' || type === '') {
+    throw new TypeError('Message type must be a non-empty string')
+  }
+  return msg
+}
+
+// ====== Client methods ====== //
+
+function connect () {
+  if (this.socket !== null) {
+    throw new Error('connect() can only be called if there is no current socket.')
+  }
+  this.socket = new WebSocket(this.url)
+
+  if (this.options.timeout) {
+    this.connectionTimeoutID = setTimeout(() => {
+      this.connectionTimeoutID = undefined
+      if (this.socket) {
+        this.socket.close(1000, 'timeout')
+      }
+    }, this.options.timeout)
+  }
+  this.reconnectionDelayID = undefined
+  // Attach WebSocket event listeners.
+  for (const name of eventNames) {
+    this.socket.addEventListener(name, this.listeners[name])
+  }
+}
+
+/**
+ * Immediately remove every event listener and free up ressources.
+ * This method is used in unit tests.
+ *
+ * - No 'close' event will be emitted.
+ * - Any incoming or outgoing buffered data will be discarded.
+ * - Any pending request will be discarded.
+ */
+function destroy () {
+  // Clear all timers.
+  clearTimeout(this.connectionTimeoutID)
+  clearTimeout(this.reconnectionDelayID)
+  // Update property values.
+  // Note: do not clear 'this.options'.
+  this.connectionTimeoutID = undefined
+  this.failedAttempts = 0
+  this.isReconnecting = false
+  this.pendingSubscriptionSet.clear()
+  this.pendingUnsubscriptionSet.clear()
+  this.reconnectionDelayID = undefined
+  this.subscriptionSet.clear()
+  // Remove global event listeners.
+  if (typeof window === 'object') {
+    for (const name of globalEventNames) {
+      window.removeEventListener(name, this.listeners[name])
+    }
+  }
+  // Remove WebSocket event listeners.
+  if (this.socket) {
+    for (const name of eventNames) {
+      this.socket.removeEventListener(name, this.listeners[name])
+    }
+    this.socket.close()
+  }
+  this.socket = null
+}
+
+function getNextRandomReconnectionDelay (): number {
+  const {
+    maxReconnectionDelay,
+    minReconnectionDelay,
+    reconnectionDelayGrowFactor
+  } = this.options
+
+  const minDelay = (
+    minReconnectionDelay * reconnectionDelayGrowFactor ** this.failedAttempts
+  )
+  const maxDelay = Math.min(
+    minReconnectionDelay * reconnectionDelayGrowFactor ** (this.failedAttempts + 1),
+    maxReconnectionDelay
+  )
+  return minDelay + Math.random() * (maxDelay - minDelay)
+}
+
+function pub (contractID: string, data: JSONType) {
+  const { socket } = this
+
+  if (socket && socket.readyState === WebSocket.OPEN) {
```

Comment:
Is it possible for the `readyState` here to not be `.OPEN`, and if so, should an error handler be added here (and elsewhere)?

Reply:
I don't know, this is just a port of the unused `pub()` implementation which was there.

A comment still mentions that `PUB` can be used for ephemeral notifications outside any contract log, but It didn't appear to be used anywhere.